### PR TITLE
SAM-2702 - Autosubmit job does not use LATEHANDLING field when retrac…

### PR DESCRIPTION
**Pull request by danielmerino**
…t date is not used.

### CODE REVIEW

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentGradingFacadeQueries.java [hunk 1]
```diff
@@ -3398,7 +3398,7 @@ public void autoSubmitAssessments() {
 						" where a.publishedAssessmentId = c.assessment.publishedAssessmentId " +
 						" and c.retractDate <= ?" +
 						" and a.status not in (5) and (a.hasAutoSubmissionRun = 0 or a.hasAutoSubmissionRun is null) and c.autoSubmit = 1 " +
-						" and a.submittedDate is not null and a.attemptDate <= c.retractDate " +
+						" and a.submittedDate is not null and (a.attemptDate <= c.retractDate or c.lateHandling = 2) " +
```
Comment (ottenhoff) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentGradingFacadeQueries.java]: This does not look correct to me.

How do you ensure that the autoSubmit job does not submit a student's *active* assessment?

Comment (danielmerino) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentGradingFacadeQueries.java]: Yes, you are right. I haven't seen that. I have done some tests and I _think_ that this alternative condition would work fine:

and (a.attemptDate <= c.retractDate or (c.dueDate < now() and c.lateHandling=2))

What do you think, Sam? If lateHandling is 2, retractDate is not used and dueDate is the field to check for autosubmit, isn't it?

The query I am using to test this in MySql while there are pending submissions is:

select a.status, a.hasAutoSubmissionRun, c.autoSubmit, c.lateHandling, assessmentgradingid, publishedassessmentid, submittedDate, attemptDate, startDate, dueDate, retractDate
from sam_AssessmentGrading_t a, sam_publishedAccessControl_t c
where a.publishedAssessmentId = c.assessmentId 
and c.retractDate <= now()
and a.status not in (5) and (a.hasAutoSubmissionRun = 0 or a.hasAutoSubmissionRun is null) and c.autoSubmit = 1 
and a.submittedDate is not null 
and (a.attemptDate <= c.retractDate or (c.dueDate < now() and c.lateHandling=2))
order by a.publishedAssessmentId, a.agentId, a.forGrade desc, a.assessmentGradingId

Comment (ottenhoff) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentGradingFacadeQueries.java]: Hi Daniel,

Yes, I think your revised SQL works. NOT_ACCEPT_LATE_SUBMISSION == 2 and that means the user cannot submit after the due date.

**Comment by ncaidin** [comment]
Does this look okay to you @ottenhoff ?

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentGradingFacadeQueries.java [hunk 2]
```diff
@@ -3398,7 +3398,8 @@ public void autoSubmitAssessments() {
 						" where a.publishedAssessmentId = c.assessment.publishedAssessmentId " +
 						" and c.retractDate <= ?" +
 						" and a.status not in (5) and (a.hasAutoSubmissionRun = 0 or a.hasAutoSubmissionRun is null) and c.autoSubmit = 1 " +
-						" and a.submittedDate is not null and a.attemptDate <= c.retractDate " +
+						" and a.submittedDate is not null " +
+						" and (a.attemptDate <= c.retractDate or (c.dueDate < now() and c.lateHandling = 2)) " +
```
Comment (ottenhoff) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentGradingFacadeQueries.java]: Please don't use "NOW()" as that provides the database server's current time and not necessarily the application server time......

**Comment by danielmerino** [comment]
Changed to use application server time as Sam suggested.

**Pull request merged**